rust/core-lib/src/tabs.rs

```diff
@@ -18,53 +18,183 @@ use std::collections::BTreeMap;
+        let buffer_id = self.views.get(view_id)
+            .expect(&format!("no buffer_id for view {}", view_id));
+        self.editors.get(buffer_id)
+    }
+
+    /// Returns a mutable reference to the `Editor` instance owning `view_id`'s view.
+    ///
+    /// Panics if no buffer is associated with `view_id`.
+    pub fn editor_for_view_mut(&mut self, view_id: &ViewIdentifier) -> Option<&mut Editor<W>> {
+        let buffer_id = self.views.get(view_id)
+            .expect(&format!("no buffer_id for view {}", view_id));
+        self.editors.get_mut(buffer_id)
+    }
+}
+
+impl <W: Write + Send + 'static>BufferContainerRef<W> {
+    pub fn new() -> Self {
+        BufferContainerRef(Arc::new(Mutex::new(
+                    BufferContainer {
+                        open_files: BTreeMap::new(),
+                        editors: BTreeMap::new(),
+                        views: BTreeMap::new(),
+                    })))
+    }
+
+    /// Returns a handle to the inner `MutexGuard`.
+    pub fn lock(&self) -> MutexGuard<BufferContainer<W>> {
+        self.0.lock()
+    }
+
+    /// Creates a new `WeakBufferContainerRef<W>`.
+    pub fn to_weak(&self) -> WeakBufferContainerRef<W> {
+        let weak_inner = Arc::downgrade(&self.0);
+        WeakBufferContainerRef(weak_inner)
+    }
+
+    /// Returns `true` if `file_path` is already open, else `false`.
+    pub fn has_open_file<P: AsRef<Path>>(&self, file_path: P) -> bool {
+        self.lock().open_files.contains_key(file_path.as_ref())
+    }
+
+    /// Adds `file_path` as an open file, associated with `view_id`'s buffer.
+    pub fn add_file<P: AsRef<Path>>(&self, file_path: P, view_id: &ViewIdentifier) {
+        let buffer_id = self.lock().views.get(view_id).unwrap().to_owned();
+        self.lock().open_files.insert(file_path.as_ref().to_owned(), buffer_id);
+    }
+
+    /// Adds a new view to the `Editor` instance owning `buffer_id`.
+    pub fn add_view(&self, view_id: &ViewIdentifier, buffer_id: &BufferIdentifier) {
+        self.lock().views.insert(view_id.to_owned(), buffer_id.to_owned());
+        self.lock().editor_for_view_mut(view_id).unwrap().add_view(view_id);
+    }
+
+    /// Closes the view with identifier `view_id`.
+    ///
+    /// If this is the last view open onto the underlying buffer, also cleans up
+    /// the `Editor` instance.
+    pub fn close_view(&self, view_id: &ViewIdentifier) {
+        let path_to_remove = {
+            let mut inner = self.lock();
+            let editor = inner.editor_for_view_mut(view_id).unwrap();
+            editor.remove_view(view_id);
+            if !editor.has_views() {
+                editor.get_path().map(PathBuf::from)
+            } else {
+            None
+            }
+        };
+
+        if let Some(path) = path_to_remove {
+            let buffer_id = self.lock().views.remove(view_id).unwrap();
+            self.lock().open_files.remove(&path);
```

Comment:
totally. Also not sure it's worth it. from "Differences from the standard library Mutex" section of the  docs for [`parking_lot::Mutex`](https://amanieu.github.io/parking_lot/parking_lot/struct.Mutex.html): 

> - Inline fast path for the uncontended case.